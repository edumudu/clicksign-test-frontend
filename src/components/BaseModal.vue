<template>
  <div
    :class="$style.wrapper"
    @click.prevent="onClose"
  >
    <div
      :class="$style.content"
      @click.stop
    >
      <div :class="$style.header">
        <slot name="header" />
      </div>

      <div :class="$style.body">
        <slot />
      </div>

      <div :class="$style.actions">
        <slot name="actions" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onBeforeUnmount } from 'vue';

const useActions = (emit: (event: 'close') => void) => {
  const onClose = () => {
    emit('close');
  };

  return {
    onClose,
  };
};

const useKeymaps = (emit: (event: 'close') => void) => {
  const { onClose } = useActions(emit);

  const keymap = {
    Escape: onClose,
    Esc: onClose,
  };

  const onKeydown = (event: KeyboardEvent) => {
    const { key } = event;

    if (!Object.keys(keymap).includes(key)) return;

    const handler = keymap[key as keyof typeof keymap];

    handler();
  };

  window.addEventListener('keydown', onKeydown);

  onBeforeUnmount(() => {
    window.removeEventListener('keydown', onKeydown);
  });
};

export default defineComponent({
  name: 'BaseModal',
  emits: ['close'],

  setup(_, { emit }) {
    useKeymaps(emit);

    return {
      ...useActions(emit),
    };
  },
});
</script>

<style lang="css" module>
.wrapper {
  position: fixed;
  inset: 0;
  background: rgba(0, 0, 0, 0.5);
  display: grid;
  place-items: center;
}

.content {
  background: var(--white);
  border-radius: 1rem;
  width: 100%;
  max-width: 432px;
}

.body {
  padding: 1.375rem 1.5rem;
  min-height: 92px;
}

.header {
  font-size: 1rem;
  color: var(--text-color);
  padding: 1rem;
  border-bottom: 2px solid var(--gray);
}

.actions {
  padding: 0.937rem 1rem;
  border-top: 2px solid var(--gray);
}
</style>
