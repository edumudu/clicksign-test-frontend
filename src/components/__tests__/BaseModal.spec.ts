import { shallowMount, VueWrapper } from '@vue/test-utils';

import BaseModal from '../BaseModal.vue';

describe('BaseModal.vue', () => {
  it('Should render content passed in default slot', async () => {
    const wrapper = shallowMount(BaseModal, {
      slots: {
        default: '<div>Test</div>',
      },
    });

    expect(wrapper.text()).toBe('Test');
  });

  it('Should render content passed in header slot', async () => {
    const wrapper = shallowMount(BaseModal, {
      slots: {
        header: '<div>Title</div>',
      },
    });

    expect(wrapper.text()).toBe('Title');
  });

  it('Should render content passed in actions slot', async () => {
    const wrapper = shallowMount(BaseModal, {
      slots: {
        actions: '<div>Actions</div>',
      },
    });

    expect(wrapper.text()).toBe('Actions');
  });

  describe('no slot', () => {
    let wrapper: VueWrapper<any>;

    beforeEach(() => {
      wrapper = shallowMount(BaseModal);
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('Should emit close event when click outside modal content', async () => {
      await wrapper.trigger('click');

      expect(wrapper.emitted('close')).toBeTruthy();
      expect(wrapper.emitted('close')).toHaveLength(1);
    });

    it('Should emit close event when press esc', async () => {
      const evt = new KeyboardEvent('keydown', { key: 'Escape' });

      window.dispatchEvent(evt);

      expect(wrapper.emitted('close')).toBeTruthy();
      expect(wrapper.emitted('close')).toHaveLength(1);
    });

    it('Should emit close event when press esc in Edge/IE', async () => {
      const evt = new KeyboardEvent('keydown', { key: 'Esc' });

      window.dispatchEvent(evt);

      expect(wrapper.emitted('close')).toBeTruthy();
      expect(wrapper.emitted('close')).toHaveLength(1);
    });
  });
});
