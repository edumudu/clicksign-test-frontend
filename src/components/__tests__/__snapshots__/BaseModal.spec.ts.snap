// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModal.vue no slot should match snapshot 1`] = `
<div
  class="wrapper"
>
  <div
    class="content"
  >
    <div
      class="header"
    >
      
      
    </div>
    <div
      class="body"
    >
      
      
    </div>
    <div
      class="actions"
    >
      
      
    </div>
  </div>
</div>
`;
